src/engine/archive.cpp

```diff
@@ -0,0 +1,125 @@
+
+        for (const auto &veh : engine.vehiclePool) {
+            const Vehicle *oldVehicle = veh.second.first;
+            Vehicle * newVehicle = new Vehicle(*oldVehicle);
+            vehiclePool.emplace(oldVehicle->getPriority(), std::make_pair(newVehicle, veh.second.second));
+        }
+        // update the vehicle pointers
+        for (const auto &veh : vehiclePool) {
+            Vehicle *vehicle = veh.second.first;
+            vehicle->laneChangeInfo.partner = getNewPointer(vehicle->laneChangeInfo.partner);
+            vehicle->controllerInfo.leader  = getNewPointer(vehicle->controllerInfo.leader);
+            vehicle->controllerInfo.blocker = getNewPointer(vehicle->controllerInfo.blocker);
+
+            std::shared_ptr<LaneChange> laneChange = vehicle->laneChange;
+            laneChange->targetLeader = getNewPointer(laneChange->targetLeader);
+            laneChange->targetFollower = getNewPointer(laneChange->targetFollower);
+            if (laneChange->signalRecv) {
+                laneChange->signalRecv = getNewPointer(laneChange->signalRecv->source)->laneChange->signalSend;
+            }
+        }
+
+        // record the information of each drivable object
+        for (const auto &drivable : engine.roadnet.getDrivables()) {
+            auto result = drivableArchive.emplace(drivable, DrivableArchive());
+            assert(result.second);
+            archiveDrivable(drivable, result.first->second);
+        }
+
+        // record the information of each flow
+        for (const auto &flow : engine.flows) {
+            auto result = flowArchive.emplace(&flow, FlowArchive());
+            assert(result.second);
+            archiveFlow(&flow, result.first->second);
+        }
+
+        //record the information of each traffic light
+        for (const auto &intersection : engine.roadnet.getIntersections()) {
+            const auto &light = &intersection.getTrafficLight();
+            auto result = trafficLightArchive.emplace(light, TrafficLightArchive());
+            assert(result.second);
+            archiveTrafficLight(light, result.first->second);
+        }
+
+    }
+
+    Vehicle *Archive::getNewPointer(const Vehicle *old) {
+        if (!old) return nullptr;
+        int priority = old->getPriority();
+        assert(vehiclePool.count(priority));
+        return vehiclePool[priority].first;
+    }
+
+    void Archive::archiveDrivable(const Drivable *drivable, Archive::DrivableArchive &archive) {
+        for (const auto &vehicle : drivable->getVehicles())
+            archive.vehicles.emplace_back(getNewPointer(vehicle));
+        if (drivable->isLane()) {
+            const Lane * lane = dynamic_cast<const Lane *>(drivable);
+            for (const auto &vehicle : lane->getWaitingBuffer()) {
+                archive.waitingBuffer.emplace_back(getNewPointer(vehicle));
+            }
+        }
+    }
+
+    void Archive::archiveFlow(const Flow *flow, Archive::FlowArchive &archive) {
```

Comment:
I prefer to rename `archive` to `flowArchive`. Name `archive` is confusing and I often think it is a variable of `class Archive`